lib/tasks/uploads.rake

```diff
@@ -666,45 +659,81 @@ def log_rebake_errors(rebake_errors)
   end
 end
 
-def update_specific_upload_security_no_login_required(upload_ids_to_mark_as_secure, upload_ids_to_mark_as_not_secure)
-  if upload_ids_to_mark_as_secure.any?
-    puts "Marking #{upload_ids_to_mark_as_secure.length} uploads as secure because UploadSecurity determined them to be secure."
-    Upload.where(id: upload_ids_to_mark_as_secure).update_all(
-      secure: true,
-      security_last_changed_at: Time.zone.now,
-      security_last_changed_reason: "upload security rake task mark as secure"
+def update_specific_upload_security_no_login_required
+  # A simplification of the rules found in UploadSecurity which is a lot faster than
+  # having to loop through records and use that class to check security.
+  post_upload_ids_marked_secure = DB.query_single(<<~SQL)
+    WITH upl AS (
+      SELECT DISTINCT ON (upload_id) upload_id
+      FROM post_uploads
+      INNER JOIN posts ON posts.id = post_uploads.post_id
+      INNER JOIN topics ON topics.id = posts.topic_id
+      LEFT JOIN categories ON categories.id = topics.category_id
+      WHERE (topics.category_id IS NOT NULL AND categories.read_restricted) OR
+        (topics.archetype = 'private_message')
     )
-  end
-  if upload_ids_to_mark_as_not_secure.any?
-    puts "Marking #{upload_ids_to_mark_as_not_secure.length} uploads as not secure because UploadSecurity determined them to be not secure."
-    Upload.where(id: upload_ids_to_mark_as_not_secure).update_all(
-      secure: false,
-      security_last_changed_at: Time.zone.now,
-      security_last_changed_reason: "upload security rake task mark as not secure"
+    UPDATE uploads
+    SET secure = true,
+        security_last_changed_reason = 'upload security rake task mark as secure',
+        security_last_changed_at = NOW()
+    FROM upl
+    WHERE uploads.id = upl.upload_id AND NOT uploads.secure
+    RETURNING uploads.id;
```

Comment:
Nitpick: I noticed you are ending all queries with `;`. I do not think that is necessary and usually we do not do it.

Reply:
Fair enough, will remove :)